File: Hw3/Main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// ASSIGNMENT 3

#define ARRAY_LENGTH 20000
#define RANDOM_RANGE 50000

// -----------------------------------------------------------------

/*
 * a)
 */

int *insertionSort (int *array, int length) {
  for (int i = 1; i < length; i++) {
    for (int y = 0; y < i; y++) {
      if (array[i] < array[y]) {
        int temp = array[y];
        array[y] = array[i];
        array[i] = temp;
      }
    }
  }
  return array;
}

/*
 * b)
 */

int *reverse (int *array, int length) {
  for (int i = 0; i < length / 2; i++) {
    int temp = array[i];
    array[i] = array[length - 1 - i];
    array[length - 1 - i] = temp;
  }
  return array;
}

// -----------------------------------------------------------------

/*
 * Helpers
 */

int *getRandomNumbers (void) {
  static int r[ARRAY_LENGTH];
  int i;
  time_t t;

  // Set the seed 
  srand((unsigned) time(&t));

  for ( i = 0; i < ARRAY_LENGTH; ++i) {
    int n = rand() / (RAND_MAX / RANDOM_RANGE + 1);
    r[i] = n;
  }
  return r;
}

float getTimeDelta (float t1, float t2) {
  return ((float)(t2 - t1) / 1000000.0F ) * 1000;
}

int isAscending (int *array, int length) {
  int isSorted = 1;
  for (int i = 0; i < length - 1; i++) {
    if (array[i] > array[i + 1]) {
      isSorted = 0;
    }
  }
  return isSorted;
}

int isDescending (int *array, int length) {
  int isSorted = 1;
  for (int i = 0; i < length - 1; i++) {
    if (array[i] < array[i + 1]) {
      isSorted = 0;
    }
  }
  return isSorted;
}

/*
 * Run this mother
 */

int main (void) {
  clock_t t1, t2;

  int *unsorted = getRandomNumbers();

  printf("\nArray of %d random numbers between 0 and %d:\n\n", 
    ARRAY_LENGTH, RANDOM_RANGE);
  
  t1 = clock();
  int *sorted = insertionSort(unsorted, ARRAY_LENGTH);
  t2 = clock();

  float diff = getTimeDelta(t1, t2);

  printf("Insertionsort:\n%f miliseconds\n\n", diff);
  printf("Array is sorted: %d\n\n", isAscending(sorted, ARRAY_LENGTH));

  t1 = clock();
  int *reversed = reverse(sorted, ARRAY_LENGTH);
  t2 = clock();

  diff = getTimeDelta(t1, t2);

  printf("Reverse:\n%f miliseconds\n\n", diff);
  printf("Array is reversed: %d\n\n", isDescending(sorted, ARRAY_LENGTH));

  return 0;
}

File: Hw3/Main.hs
module Main where

import Control.Monad
import Data.Time.Clock
import System.CPUTime
import System.Random
import Data.List (insert)

-- ASSIGNMENT 3

listLength = 20000
randomRange = 50000

-------------------------------------------------------------------

-- a)

insertionSort :: [Int] -> [Int]
insertionSort = foldr insert []

-- b)

reverse' :: [Int] -> [Int]
reverse' = foldl (flip (:)) []

-------------------------------------------------------------------

-- Helpers

getTimeDelta :: UTCTime -> UTCTime -> Float
getTimeDelta t1 t2 = fromRational . ((*) 1000) . toRational $ utctDayTime t2 - utctDayTime t1

isAscending :: [Int] -> Bool
isAscending [] = False
isAscending [x] = True
isAscending [x, y] = x <= y
isAscending (x:y:ys) = x <= y && isAscending (y:ys)

isDescending:: [Int] -> Bool
isDescending [] = False
isDescending [x] = True
isDescending [x, y] = x >= y
isDescending (x:y:ys) = x >= y && isDescending (y:ys)

main = do
  unsorted <- replicateM listLength $ randomRIO (0, randomRange)

  putStrLn $ "List of " ++ show listLength ++ 
             " random numbers between 0 and " ++ show randomRange ++ ":"

  before <- getCurrentTime
  sorted <- return $! insertionSort unsorted
  after <- getCurrentTime

  let timeDelta = getTimeDelta before after

  putStrLn $ "insertionSort: " ++ show timeDelta ++ "ms"
  putStrLn $ "List is sorted: " ++ show (isAscending sorted)

  before' <- getCurrentTime
  reversed <- return $! reverse' sorted
  after' <- getCurrentTime

  let timeDelta' = getTimeDelta before' after'

  putStrLn $ "reverse: " ++ show timeDelta' ++ "ms"
  putStrLn $ "List is reversed: " ++ show (isDescending reversed)

